spark3-extensions/src/main/scala/org/apache/spark/sql/catalyst/optimizer/RewriteDelete.scala

```diff
@@ -0,0 +1,160 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.spark.sql.catalyst.optimizer
+
+import java.util.UUID
+import org.apache.spark.sql.{sources, AnalysisException}
+import org.apache.spark.sql.catalyst.analysis.Resolver
+import org.apache.spark.sql.catalyst.expressions.{Alias, Attribute, AttributeReference, EqualNullSafe, Expression, InputFileName, Literal, Not, PredicateHelper, SubqueryExpression}
+import org.apache.spark.sql.catalyst.expressions.Literal.FalseLiteral
+import org.apache.spark.sql.catalyst.plans.logical.{Aggregate, DeleteFromTable, DynamicFileFilter, Filter, LocalRelation, LogicalPlan, OverwriteFiles, Project}
+import org.apache.spark.sql.catalyst.rules.Rule
+import org.apache.spark.sql.connector.catalog.{SupportsMetadataOnlyDeletes, Table}
+import org.apache.spark.sql.connector.write.{LogicalWriteInfo, LogicalWriteInfoImpl, RowLevelOperationsBuilder, SupportsPushDownRowLevelFilters}
+import org.apache.spark.sql.execution.datasources.DataSourceStrategy
+import org.apache.spark.sql.execution.datasources.v2.{DataSourceV2Relation, DataSourceV2ScanRelation}
+import org.apache.spark.sql.internal.SQLConf
+import org.apache.spark.sql.types.{BooleanType, StructType}
+import org.apache.spark.sql.util.CaseInsensitiveStringMap
+
+// TODO: should be part of early scan push down after the delete condition is optimized
+object RewriteDelete extends Rule[LogicalPlan] with PredicateHelper {
+
+  import org.apache.spark.sql.execution.datasources.v2.ExtendedDataSourceV2Implicits._
+
+  private val FILE_NAME_COL = "_file"
+
+  override def apply(plan: LogicalPlan): LogicalPlan = plan resolveOperators {
+    // no need to execute this delete as the condition is false
+    case DeleteFromTable(_, Some(FalseLiteral)) =>
+      LocalRelation()
```

Comment:
I think this should be done in `run` instead of replacing the plan with a no-op. That way it is clear that the plan did run, it just didn't have an effect.

Reply:
I agree but I am not sure we can modify the exec node for deletes.

Reply:
Oh, I see. If this is intended to avoid this getting rewritten to a different plan, then let's add a comment to it?

Reply:
Well, you are right, it is weird. I think we can get rid of this branch at all. It should be covered by `deleteWhere` case. We just need to adapt our delete with filters implementation to handle `false` predicates.

Reply:
Do we need to change anything? Currently, it should try to generate a commit, but not remove anything and probably not even write new metadata.

Reply:
I think we just need to return early instead of trying the delete and produce a new snapshot without changes.